File: src/main/java/com/github/tianchenghang/tools/TerminateTool.java
package com.github.tianchenghang.tools;

import org.springframework.ai.tool.annotation.Tool;

/** Termination tool (allows autonomous planning agents to gracefully interrupt execution). */
public class TerminateTool {

  @Tool(
      description =
          """
      Terminate the interaction when the request is met or if the assistant cannot proceed further with the task.
      Call this tool when all tasks are completed to end the session.
      """)
  public String doTerminate() {
    return "Task completed";
  }
}

File: server/src/tools/terminate-tool.ts
import { StructuredTool } from '@langchain/core/tools';
import { z } from 'zod';

export class TerminateTool extends StructuredTool {
  name = TerminateTool.name;
  description = `
    Terminate the interaction when the request is met or if the assistant cannot proceed further with the task.
    Call this tool when all tasks are completed to end the session.
  `;
  schema = z.object({});

  async _call(): Promise<string> {
    return 'Task completed';
  }
}
